planner/core/find_best_task.go

```diff
@@ -424,8 +424,12 @@ func (ds *DataSource) convertToIndexScan(prop *property.PhysicalProperty, path *
 // TODO: refactor this part, we should not call Clone in fact.
 func (is *PhysicalIndexScan) initSchema(id int, idx *model.IndexInfo, isDoubleRead bool) {
 	indexCols := make([]*expression.Column, 0, len(idx.Columns))
+	baseSchema := expression.TableInfo2SchemaWithDBName(is.ctx, is.DBName, is.Table)
```

Comment:
The columns in `datasourceSchema` is generated in the same as `TableInfo2SchemaWithDBName`.
Why this change can take effect?

Reply:
Let's consider the table and the query below:
```
use test;
create table t(a int, b int, c int, primary key (a, b));
prepare stmt1 from "select a, c from t where a = ? and c = ?";
set @a=1, @c=3;
execute stmt1 using @a, @c;
```
Before the PR, ``datasourceSchema`` contains only two columns, ``test.t.a`` and ``test.t.c``. However, ``indexCols``  contains ``test.t.a``, ``b``, ``_row_id``. Here, the ``RetType`` of the ``Colume`` expression of ``b`` is ``nil`` because ``b`` is not in ``datasourceSchema``. It causes the panic when accessing the ``RetType`` of ``b``

After the PR, ``indexCols``  contains ``test.t.a``, ``test.t.b``, ``_row_id``, and the ``RetType`` of the ``Colume`` expression of ``b`` is not ``nil``.

Reply:
Oh, i got it.
Seems that you can directly use `PhysicalIndexScan.Table.Columns[IndexCol.Offset].FieldType` to get its `RetType`.

Reply:
Yes, it should be enough.